### DISCUSSION TITLE
&T and &T mut disjoint sets question

### DISCUSSION THREAD

**Discussion by jasonheath** [pretzelhammer/rust-blog#65] [General]
I'd like to ask where I can read more about about the &T and &mut T disjoint sets mentioned in the first misconception in ([T only contains owned types](https://github.com/pretzelhammer/rust-blog/blob/master/posts/common-rust-lifetime-misconceptions.md#1-t-only-contains-owned-types)) in your [Common Rust Lifetime Conceptions](https://github.com/pretzelhammer/rust-blog/blob/master/posts/common-rust-lifetime-misconceptions.md) blog post. I think I get the essential idea of the first point. When it comes to generics T is the set of "everything" as opposed to "just owned types" as the T, &T, and &mut T data types might trick one into believing. &T and &mut T are disjoint sets containing all the variants of immutable and mutable references. Where I get confused is in trying to understand how the two disjoint sets grow, what's the pattern?

For example, "&&mut i32" in the &T examples makes sense if I read that as "an immutable reference to a mutable reference." Similarly, "&mut &i32" makes sense as a "mutable reference to an immutable reference". But I'm not as clear as I'd like as to how those two sets permute as they grow.  I haven't seen anyone ask similar in any of the discussions and nothing I've found so far online has cleared it up for me so I'd thought I'd ask for pointers toward additional reading.

**Comment by pretzelhammer** [comment]
Well, it boils down to `T` being "everything" and if you put a `&` in front of `T` to get `&T` which is just "everything but with a `&` in front of it" and if you put a `&mut` in front of `T` you get `&mut T` which is just "everything but with a `&mut` in front of it".

Also, I wouldn't say any of these abstract sets "grow", they'll all already just infinite. You can take any Rust type `T` and make it a reference with `&` to make `&T`, but `&T` itself is a Rust type, so you can repeatedly apply this operation ad infinitum, e.g. `&T` -> `&&T` -> `&&&T` -> `&&&&T` and so on until you have infinitely many types in your set. You can do this similarly with `&mut` and with a mix of `&` and `&mut`.

I hope that explanation helps clarify some things!

**Reply by jasonheath** [reply]
Thanks @pretzelhammer!
